### PR TITLE
command/init: Be explicit that some options are not relevant for Terraform Cloud mode

### PR DESCRIPTION
**Pull request by apparentlymart**
There are a few command line options for `terraform init` which are only relevant when working with traditional backends, with the Cloud integration previously just mostly ignoring them, or sometimes misbehaving slightly due to them creating an unreasonable situation.

Now we'll catch these and return explicit errors, in order to be clear that these options are not needed nor supported in Cloud mode.

### CODE REVIEW

**Comment by apparentlymart** [comment]
As of now, this has the following e2etest failures that I still need to work through and understand whether these behavior changes are an expected result of the simplification this PR implies or if they represent regressions.

```
--- FAIL: Test_migrate_remote_backend_name_to_tfc_name (33.13s)
    migrate_state_remote_backend_to_tfc_test.go:118: read |0: i/o timeout
--- FAIL: Test_migrate_remote_backend_name_to_tfc_same_name (34.80s)
    migrate_state_remote_backend_to_tfc_test.go:221: error while waiting for output
        want: Terraform Cloud has been successfully initialized!
        error: read |0: i/o timeout
        output
         Terraform will automatically
        use this backend unless the backend configuration changes.
        
        Initializing provider plugins...
        
        Terraform has been successfully initialized!
        
        You may now begin working with Terraform. Try running "terraform plan" to see
        any changes that are required for your infrastructure. All Terraform commands
        should now work.
        
        If you ever set or change modules or backend configuration for Terraform,
        rerun this command to reinitialize your working directory. If you forget, other
        commands will detect it and remind you to do so if necessary.
        Running apply in the remote backend. Output will stream here. Pressing Ctrl-C
        will cancel the remote apply if it's still pending. If the apply started it
        will stop streaming the logs, but will not stop the apply running remotely.
        
        Preparing the remote apply...
        
        To view this run in a browser, visit:
        https://app.staging.terraform.io/app/tst-c5b1aca1-4604-8edb-b28b-11e4ad4b44ad/remote-workspace/runs/run-sthS8wUBFeJSmix5
        
        Waiting for the plan to start...
        
        Terraform v1.0.11
        on linux_amd64
        Configuring remote state backend...
        Initializing Terraform configuration...
        
        Changes to Outputs:
          + val = "default"
        
        You can apply this plan to save these new output values to the Terraform
        state, without changing any real infrastructure.
        
        
        Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
        
        Outputs:
        
        val = "default"
        
        Initializing Terraform Cloud...
        Migrating from backend "remote" to Terraform Cloud.
        Do you wish to proceed?
          As part of migrating to Terraform Cloud, Terraform can optionally copy your
          current workspace state to the configured Terraform Cloud workspace.
          
          Answer "yes" to copy the latest state snapshot to the configured
          Terraform Cloud workspace.
          
          Answer "no" to ignore the existing state and just activate the configured
          Terraform Cloud workspace with its existing state, if any.
          
          Should Terraform migrate your existing state?
        
          Enter a value: 
--- FAIL: Test_migrate_remote_backend_name_to_tfc_name_different_org (45.58s)
    migrate_state_remote_backend_to_tfc_test.go:364: read |0: i/o timeout
--- FAIL: Test_migrate_remote_backend_name_to_tfc_tags (33.93s)
    migrate_state_remote_backend_to_tfc_test.go:481: error while waiting for output
        want: Terraform Cloud requires all workspaces to be given an explicit name.
        error: read |0: i/o timeout
        output
        "
        
        You can apply this plan to save these new output values to the Terraform
        state, without changing any real infrastructure.
        
        
        Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
        
        Outputs:
        
        val = "default"
        default
        
        Initializing Terraform Cloud...
        Migrating from backend "remote" to Terraform Cloud.
        Do you wish to proceed?
          As part of migrating to Terraform Cloud, Terraform can optionally copy your
          current workspace state to the configured Terraform Cloud workspace.
          
          Answer "yes" to copy the latest state snapshot to the configured
          Terraform Cloud workspace.
          
          Answer "no" to ignore the existing state and just activate the configured
          Terraform Cloud workspace with its existing state, if any.
          
          Should Terraform migrate your existing state?
        
          Enter a value: 
--- FAIL: Test_migrate_remote_backend_prefix_to_tfc_name_strategy_single_workspace (38.57s)
    migrate_state_remote_backend_to_tfc_test.go:621: error while waiting for output
        want: Terraform Cloud has been successfully initialized!
        error: read |0: i/o timeout
        output
        
        
          Enter a value: yes
        
        Do you want to copy existing state to Terraform Cloud?
          Pre-existing state was found while migrating the previous "remote" backend to Terraform Cloud.
          No existing state was found in Terraform Cloud. Do you want to copy this state to Terraform Cloud?
          Enter "yes" to copy and "no" to start with an empty state.
        
          Enter a value: 
--- FAIL: Test_migrate_remote_backend_prefix_to_tfc_tags_strategy_single_workspace (42.04s)
    migrate_state_remote_backend_to_tfc_test.go:884: error while waiting for output
        want: Terraform Cloud requires all workspaces to be given an explicit name.
        error: read |0: i/o timeout
        output
        
        
        You may now begin working with Terraform. Try running "terraform plan" to see                                                                                    
        any changes that are required for your infrastructure. All Terraform commands                                                                                    
        should now work.                                                                                                                                                 
                                                                                                                                                                         
        If you ever set or change modules or backend configuration for Terraform,                                                                                        
        rerun this command to reinitialize your working directory. If you forget, other                                                                                  
        commands will detect it and remind you to do so if necessary.                                                                                                    
        Running apply in the remote backend. Output will stream here. Pressing Ctrl-C
        will cancel the remote apply if it's still pending. If the apply started it                                                                                      
        will stop streaming the logs, but will not stop the apply running remotely.                                                                                      
        
        Preparing the remote apply...
        
        To view this run in a browser, visit:
        https://app.staging.terraform.io/app/tst-c352cd05-a50f-d19c-c34f-9152eacb1721/app-one/runs/run-N8DfBG6zCHfALDHU                                                  
        
        Waiting for the plan to start...
        
        Terraform v1.0.11
        on linux_amd64
        Configuring remote state backend...
        Initializing Terraform configuration...
        
        Changes to Outputs:
          + val = "default"
        
        You can apply this plan to save these new output values to the Terraform
        state, without changing any real infrastructure.
        
        
        Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
        
        Outputs:
        
        val = "default"
        
        Initializing Terraform Cloud...
        Migrating from backend "remote" to Terraform Cloud.
        Do you wish to proceed?
          As part of migrating to Terraform Cloud, Terraform can optionally copy your
          current workspace state to the configured Terraform Cloud workspace.
          
          Answer "yes" to copy the latest state snapshot to the configured
          Terraform Cloud workspace.
          
          Answer "no" to ignore the existing state and just activate the configured
          Terraform Cloud workspace with its existing state, if any.
          
          Should Terraform migrate your existing state?
        
          Enter a value: 
--- FAIL: Test_migrate_single_to_tfc (0.22s)
    --- FAIL: Test_migrate_single_to_tfc/migrate_using_cloud_workspace_tags_strategy (16.47s)
        migrate_state_single_to_tfc_test.go:166: error while waiting for output
            want: Terraform Cloud requires all workspaces to be given an explicit name.
            error: read |0: i/o timeout
            output
             Terraform will automatically
            use this backend unless the backend configuration changes.
            
            Initializing provider plugins...
            
            Terraform has been successfully initialized!
            
            You may now begin working with Terraform. Try running "terraform plan" to see                                                                                
            any changes that are required for your infrastructure. All Terraform commands                                                                                
            should now work.                                                                                                                                             
                                                                                                                                                                         
            If you ever set or change modules or backend configuration for Terraform,                                                                                    
            rerun this command to reinitialize your working directory. If you forget, other                                                                              
            commands will detect it and remind you to do so if necessary.                                                                                                
            
            Changes to Outputs:
              + val = "default"
            
            You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.
            
            Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
            
            Outputs:
            
            val = "default"
            
            Initializing Terraform Cloud...
            Migrating from backend "local" to Terraform Cloud.
            Do you wish to proceed?
              As part of migrating to Terraform Cloud, Terraform can optionally copy your
              current workspace state to the configured Terraform Cloud workspace.
              
              Answer "yes" to copy the latest state snapshot to the configured
              Terraform Cloud workspace.
              
              Answer "no" to ignore the existing state and just activate the configured
              Terraform Cloud workspace with its existing state, if any.
              
              Should Terraform migrate your existing state?
            
              Enter a value: 
    --- FAIL: Test_migrate_single_to_tfc/migrate_using_cloud_workspace_name_strategy (16.86s)
        migrate_state_single_to_tfc_test.go:182: read |0: i/o timeout
--- FAIL: Test_migrate_tfc_to_tfc_single_workspace (95.16s)
    --- FAIL: Test_migrate_tfc_to_tfc_single_workspace/migrating_from_name_to_name (31.24s)
        migrate_state_tfc_to_tfc_test.go:257: error while waiting for output
            want: Should Terraform migrate your existing state?
            error: read |0: i/o timeout
            output
            
            Running apply in Terraform Cloud. Output will stream here. Pressing Ctrl-C
            will cancel the remote apply if it's still pending. If the apply started it                                                                                  
            will stop streaming the logs, but will not stop the apply running remotely.                                                                                  
            
            Preparing the remote apply...
            
            To view this run in a browser, visit:
            https://app.staging.terraform.io/app/tst-18b53e8a-e603-27b6-6777-ad1792295974/prod/runs/run-F7LCDTbwyq2Eh7H3                                                 
            
            Waiting for the plan to start...
            
            Terraform v1.1.0-beta1
            on linux_amd64
            Configuring remote state backend...
            Initializing Terraform configuration...
            
            Changes to Outputs:
              + val = "prod"
            
            You can apply this plan to save these new output values to the Terraform
            state, without changing any real infrastructure.
            
            
            Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
            
            Outputs:
            
            val = "prod"
            
            Initializing Terraform Cloud...
            Terraform Cloud configuration has changed.
            
            Initializing provider plugins...
            
            Terraform Cloud has been successfully initialized!
            
            You may now begin working with Terraform Cloud. Try running "terraform plan" to                                                                              
            see any changes that are required for your infrastructure.                                                                                                   
                                                                                                                                                                         
            If you ever set or change modules or Terraform Settings, run "terraform init"                                                                                
            again to reinitialize your working directory.                                                                                                                
    --- FAIL: Test_migrate_tfc_to_tfc_single_workspace/migrating_from_name_to_tags (33.48s)
        migrate_state_tfc_to_tfc_test.go:257: error while waiting for output
            want: Terraform Cloud requires all workspaces to be given an explicit name.
            error: read |0: i/o timeout
            output
            
            
            You may now begin working with Terraform Cloud. Try running "terraform plan" to                                                                              
            see any changes that are required for your infrastructure.                                                                                                   
                                                                                                                                                                         
            If you ever set or change modules or Terraform Settings, run "terraform init"                                                                                
            again to reinitialize your working directory.                                                                                                                
            Running apply in Terraform Cloud. Output will stream here. Pressing Ctrl-C
            will cancel the remote apply if it's still pending. If the apply started it                                                                                  
            will stop streaming the logs, but will not stop the apply running remotely.                                                                                  
            
            Preparing the remote apply...
            
            To view this run in a browser, visit:
            https://app.staging.terraform.io/app/tst-5198a511-8374-2f0a-ce12-6e7ed229aafd/prod/runs/run-KM4agpYTneAC4xyi                                                 
            
            Waiting for the plan to start...
            
            Terraform v1.1.0-beta1
            on linux_amd64
            Configuring remote state backend...
            Initializing Terraform configuration...
            
            Changes to Outputs:
              + val = "prod"
            
            You can apply this plan to save these new output values to the Terraform
            state, without changing any real infrastructure.
            
            
            Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
            
            Outputs:
            
            val = "prod"
            
            Initializing Terraform Cloud...
            Terraform Cloud configuration has changed.
            
            No workspaces found.
              
              There are no workspaces with the configured tags (app)
              in your Terraform Cloud organization. To finish initializing, Terraform needs at
              least one workspace available.
              
              Terraform can create a properly tagged workspace for you now. Please enter a
              name to create a new Terraform Cloud workspace.
            
              Enter a value: 
    --- FAIL: Test_migrate_tfc_to_tfc_single_workspace/migrating_from_name_to_tags_without_ignore-version_flag (30.20s)
        migrate_state_tfc_to_tfc_test.go:257: error while waiting for output
            want: Terraform Cloud requires all workspaces to be given an explicit name.
            error: read |0: i/o timeout
            output
            
            
            You may now begin working with Terraform Cloud. Try running "terraform plan" to                                                                              
            see any changes that are required for your infrastructure.                                                                                                   
                                                                                                                                                                         
            If you ever set or change modules or Terraform Settings, run "terraform init"                                                                                
            again to reinitialize your working directory.                                                                                                                
            Running apply in Terraform Cloud. Output will stream here. Pressing Ctrl-C
            will cancel the remote apply if it's still pending. If the apply started it                                                                                  
            will stop streaming the logs, but will not stop the apply running remotely.                                                                                  
            
            Preparing the remote apply...
            
            To view this run in a browser, visit:
            https://app.staging.terraform.io/app/tst-528e4a6a-a1fb-7909-2160-39add78bb658/prod/runs/run-E4pzWuhP5AXtS31q                                                 
            
            Waiting for the plan to start...
            
            Terraform v1.1.0-beta1
            on linux_amd64
            Configuring remote state backend...
            Initializing Terraform configuration...
            
            Changes to Outputs:
              + val = "prod"
            
            You can apply this plan to save these new output values to the Terraform
            state, without changing any real infrastructure.
            
            
            Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
            
            Outputs:
            
            val = "prod"
            
            Initializing Terraform Cloud...
            Terraform Cloud configuration has changed.
            
            No workspaces found.
              
              There are no workspaces with the configured tags (app)
              in your Terraform Cloud organization. To finish initializing, Terraform needs at
              least one workspace available.
              
              Terraform can create a properly tagged workspace for you now. Please enter a
              name to create a new Terraform Cloud workspace.
            
              Enter a value: 
--- FAIL: Test_migrate_tfc_to_tfc_multiple_workspace (83.86s)
    --- FAIL: Test_migrate_tfc_to_tfc_multiple_workspace/migrating_from_multiple_workspaces_via_tags_to_name (42.26s)
        migrate_state_tfc_to_tfc_test.go:533: read |0: i/o timeout
        helper_test.go:69: Error destroying organization! WARNING: Dangling resources
            may exist! The full error is shown below.
            
            Organization: tst-e193f19f-53e8-9299-0c2c-4a77f0079c40
            Error: internal server error
    --- FAIL: Test_migrate_tfc_to_tfc_multiple_workspace/migrating_from_multiple_workspaces_via_tags_to_other_tags (41.24s)
        migrate_state_tfc_to_tfc_test.go:517: error while waiting for output
            want: Would you like to rename your workspaces?
            error: read |0: i/o timeout
            output
             Resources: 0 added, 0 changed, 0 destroyed.
            
            Outputs:
            
            tag_val = "app"
            
            Initializing Terraform Cloud...
            Terraform Cloud configuration has changed.
            
            No workspaces found.
              
              There are no workspaces with the configured tags (billing)
              in your Terraform Cloud organization. To finish initializing, Terraform needs at
              least one workspace available.
              
              Terraform can create a properly tagged workspace for you now. Please enter a
              name to create a new Terraform Cloud workspace.
            
              Enter a value: 
```

**Comment by barrettclark** [comment]
I've got the e2e tests passing now

**Comment by apparentlymart** [comment]
Thanks @barrettclark! I'm gonna retroactively fix the tests I broke while I was cleaning up the error messages in an earlier commit, and to make it so the (non-e2e) tests are passing at all of the commits here I intend to force-push to update that earlier commit where I neglected to update those, and so I'll inevitably be rewriting the e2e test commit at the same time.

File: internal/cloud/migration.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+	switch {
+	case configIsCloud:
+		switch {
+		case wdIsCloud || wdIsUninit:
+			// If config has cloud and the working directory is completely
+			// uninitialized then we assume we're doing the initial activation
+			// of this working directory for an already-migrated-to-cloud
+			// remote state.
+			return ConfigChangeInPlace
```
[the remark this answers is not in the document]
Comment (chrisarcand) [reply internal/cloud/migration.go:70]: (This isn't worth editing at the moment though, tbh. Just me musing.)

File: internal/cloud/configchangemode_string.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	case i == 8600:
+		return _ConfigChangeMode_name_1
+	case i == 8635:
+		return _ConfigChangeMode_name_2
+	case i == 129335:
+		return _ConfigChangeMode_name_3
+	default:
+		return "ConfigChangeMode(" + strconv.FormatInt(int64(i), 10) + ")"
+	}
+}
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/cloud/configchangemode_string.go:37]: The file you've commented on here is generated by the standard Go tool `stringer` and is something we typically do for all "enum-like" types in Terraform.

The use of rune is more of a Terraform-specific quirk that I was following here, and I believe the reason for it is something subtle like that it reduces the number situations where the Go compiler will allow an implicit conversion from a plain `int`, and thus catches at compile time some mistakes that we historically only caught at runtime. My proximal motivation here was just consistency with how we usually do things though, to be honest, rather than any reason specific to this situation.

File: internal/cloud/migration.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/cloud/migration.go:70]: This case has, I suppose, actually ended up incorporating two separate cases I had on my first pass:
* User previously run `terraform init` with a `cloud` block, and has now changed the `cloud` block contents and run `terraform init` again. That's what "in place" was intended to mean here.
* User has not previously initialized this working directory _at all_, but the configuration already contains a `cloud` block and so `terraform init` should behave as if we were already using Terraform Cloud, rather than proposing any migration. I originally had a separate case for this but later noticed that distinguishing between "still using cloud, config unchanged" and "still using cloud, config changed" is handled elsewhere in the backend init control flow and so melded them together into a single case, which I can see is now perhaps not suitably named for its new role.

A "Rename Symbol" should make fast work of changing this later, so let's let this sit for now and think separately about what might be better to call thus.

File: internal/command/meta_backend.go [hunk 3]
```diff
@@ -1376,17 +1477,26 @@ hasn't changed and try again. At this point, no changes to your existing
...
+
+To re-initialize, run:
+  terraform init
+
+Terraform has not yet made changes to your existing configuration or state.
+`
+
 const errBackendInitCloud = `
-Reason: %s
+Reason: %s.
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/command/meta_backend.go:1492]: This was an "while I'm in the area" error message improvement: diagnostic detail messages are always full paragraphs with punctuation, and this was only missing its punctuation because it was originally (prior to v0.12) not a diagnostic message at all.

The motivation here is the same as removing the hard line breaks in some of the other messages: gradually making these old codepaths more consistent with our modern expectations while we are working in the area anyway.

File: internal/command/meta_backend_migrate.go [hunk 4]
```diff
@@ -569,10 +569,20 @@ func (m *Meta) backendMigrateTFC(opts *backendMigrateOpts) error {
...
 		// strategy.
+		//
+		// We do prompt first though, because state migration is mandatory
+		// for moving to Cloud and the user should get an opportunity to
+		// confirm that first.
+		if migrate, err := m.promptSingleToCloudSingleStateMigration(opts); err != nil {
+			return err
+		} else if !migrate {
+			return nil //skip migrating but return successfully
+		}
```
[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/command/meta_backend_migrate.go:584]: Honestly, I'm not totally sure... when I first tested this I found some paths that were previously not prompting at all, and that's why I added this here. It looks like the other function belongs to a different migration function that handles some different cases, but I agree we can probably consolidate these two functions together if their error messages are equivalent enough.

Can we hold this for a later change, since it seems like it'll just be a pretty trivial in-place swap to the other function and then delete the redundant one I added here?

[the remark this answers is not in the document]
Comment (apparentlymart) [reply internal/command/meta_backend_migrate.go:594]: Follow-up: I added an extra commit to avoid printing out the redundant migration message this implies in some paths, so I'm hoping that's a reasonable _immediate_ solution which we can polish in a later commit.

**Review by alisdair — APPROVED** [review]
From reviewing the diff only (not local testing), this makes sense to me.

**Review by chrisarcand — APPROVED** [review]
From local testing, and not digging too deep on the diff, this makes sense to me 👍🏻

There's several bits around copy that could be clearer and maybe some identifiers in the code that could be renamed, but nothing that merits holding up merging right now.

The cloud E2E's may or may not be passing at this point, we don't have utmost confidence that any failures are related to these changes are not, so will not delay merging on behalf of those, either.

In other words, with @barrettclark 's approval (and any last E2E changes), I think we're good ✅

**Review by barrettclark — APPROVED** [review]
